### PR TITLE
feat: Text Utils

### PR DESCRIPTION
**Pull request by AlbertUnruh**
## What type of pull request is this?

- [x] Non-breaking code change
- [ ] Breaking code change
- [ ] Documentation change/addition
- [ ] Tests change

## Description
Adds a function to check whether something was mentioned somewhere.

## Changes

- add `naff.client.utils.text_utils`

## Checklist

- [x] I've formatted my code with [Black](https://black.readthedocs.io/en/stable/)
- [x] I've ensured my code works on `Python 3.10.x`
- [x] I've tested my code

### CODE REVIEW

File: naff/models/discord/message.py [hunk 1]
```diff
@@ -380,6 +381,15 @@ def get_referenced_message(self) -> Optional["Message"]:
             return None
         return self._client.cache.get_message(self._channel_id, self._referenced_message_id)
 
+    def does_mention(self, query: "str | re.Pattern[str] | models.BaseUser | models.BaseChannel | models.Role") -> bool:
```
Comment (AlbertUnruh) [inline naff/models/discord/message.py]: *I'm open for other names for the method

Comment (silasary) [reply naff/models/discord/message.py]: Let's call it `contains_mention`

File: naff/client/utils/text_utils.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+def mentions(text: str, query: "str | re.Pattern[str] | models.BaseUser | models.BaseChannel | models.Role") -> bool:
+    """Checks whether a query is present in a text."""
+    # don't use match/case witch type(query) since subclasses aren't checked
+    if isinstance(query, str):
+        return query in text
+    elif isinstance(query, re.Pattern):
+        return query.match(text) is not None
+    elif isinstance(query, models.BaseUser):
+        # mentions with <@!ID> aren't detected without the replacement
+        return (query.mention in text.replace("@!", "@")) or (query.username in text)
```
Comment (LordOfPolls) [inline naff/client/utils/text_utils.py]: Why are you checking if the username is in the text? `LordOfPolls` isn't a mention `<@174918559539920897>` is

Further, if you removed that superfluous check you could squash this and the below if statements into one

Comment (AlbertUnruh) [reply naff/client/utils/text_utils.py]: I don't want to check for (Discord) mentions, but for the presence itself (if it makes sense).

File: naff/models/discord/message.py (hunk 1 above)
Comment (AlbertUnruh) [reply naff/models/discord/message.py]: I'll change it as soon I'm on my computer

File: naff/client/utils/text_utils.py (hunk 2 above)
Comment (LordOfPolls) [reply naff/client/utils/text_utils.py]: That seems counter-intuitive to the feature's name, nor how most people would assume it works. 

Assuming we did go that route, however, using `display_name` would be better surely.

File: naff/client/utils/text_utils.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import re
+import naff.models as models
+
+__all__ = ("mentions",)
+
+
+def mentions(text: str, query: "str | re.Pattern[str] | models.BaseUser | models.BaseChannel | models.Role") -> bool:
+    """Checks whether a query is present in a text."""
+    # don't use match/case witch type(query) since subclasses aren't checked
```
Comment (LordOfPolls) [inline naff/client/utils/text_utils.py]: You need to add the parameters to the docstring

File: naff/client/utils/text_utils.py (hunk 2 above)
Comment (AlbertUnruh) [reply naff/client/utils/text_utils.py]: *at least I would check for `BaseUser.tag` -> `LordOfPolls#1010`

File: naff/client/utils/text_utils.py (hunk 3 above)
Comment (AlbertUnruh) [reply naff/client/utils/text_utils.py]: I'll add it later *(I'm out for the day, should have time in the evening (UTC+2))*

File: naff/client/utils/text_utils.py (hunk 2 above)
Comment (AlbertUnruh) [reply naff/client/utils/text_utils.py]: > That seems counter-intuitive to the feature's name, nor how most people would assume it works. 
> 
> Assuming we did go that route, however, using `display_name` would be better surely. 

Any other suggestions for the features name?

Comment (LordOfPolls) [reply naff/client/utils/text_utils.py]: I would suggest removing the username detection and keeping the name as it is. ;)

Comment (silasary) [reply naff/client/utils/text_utils.py]: Username detection really feels prone to false positives.  I definitely wouldn't want it by default.

File: naff/client/utils/text_utils.py (hunk 3 above)
Comment (AlbertUnruh) [reply naff/client/utils/text_utils.py]: added with [728bb03](https://github.com/NAFTeam/NAFF/pull/700/commits/728bb038c0bec1c8d7f0a680965238a9089d489e)

File: naff/client/utils/text_utils.py (hunk 2 above)
Comment (AlbertUnruh) [reply naff/client/utils/text_utils.py]: like so?
https://github.com/NAFTeam/NAFF/blob/da514aa1f91f422a55224fd7ba0da322b5a90df9/naff/client/utils/text_utils.py#L23-L25

File: naff/client/const.py [hunk 4]
```diff
@@ -96,6 +98,7 @@ def get_logger() -> logging.Logger:
 
 default_locale = "english_us"
 kwarg_spam = False
+tag_as_mention = True
```
Comment (silasary) [inline naff/client/const.py]: Why is this in const?
It should:
A. Be a parameter on contains_mention
B. Be False by default.

Comment (AlbertUnruh) [reply naff/client/const.py]: Well... I was tired yesterday

I'll change it later

Comment (AlbertUnruh) [reply naff/client/const.py]: changed with [03e2f64](https://github.com/NAFTeam/NAFF/pull/700/commits/03e2f64f2de76eca3749fa40cc8dad285ed7789b)

File: naff/client/utils/text_utils.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    query: "str | re.Pattern[str] | models.BaseUser | models.BaseChannel | models.Role",
+    *,
+    tag_as_mention: bool = False,
+) -> bool:
+    """Checks whether a query is present in a text.
+
+    Args:
+        text: The text to search in
+        query: The query to search for
+        tag_as_mention: Should ``BaseUser.tag`` be checked *(only if query is an instance of BaseUser)*
```
Comment (LordOfPolls) [inline naff/client/utils/text_utils.py]:
```suggestion
        tag_as_mention: Should `BaseUser.tag` be checked *(only if query is an instance of BaseUser)*
```

File: naff/models/discord/message.py [hunk 6]
```diff
@@ -380,6 +381,24 @@ def get_referenced_message(self) -> Optional["Message"]:
...
+        query: "str | re.Pattern[str] | models.BaseUser | models.BaseChannel | models.Role",
+        *,
+        tag_as_mention: bool = False,
+    ) -> bool:
+        """
+        Check whether the message contains the query or not.
+
+        Args:
+            query: The query to search for
+            tag_as_mention: Should ``BaseUser.tag`` be checked *(only if query is an instance of BaseUser)*
```
Comment (LordOfPolls) [inline naff/models/discord/message.py]:
```suggestion
            tag_as_mention: Should `BaseUser.tag` be checked *(only if query is an instance of BaseUser)*
```

File: naff/client/utils/text_utils.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    Returns:
+        Whether the query could be found in the text
+    """
+    # don't use match/case witch type(query) since subclasses aren't checked
```
Comment (LordOfPolls) [inline naff/client/utils/text_utils.py]: Redundant 
```suggestion
```

**Review by LordOfPolls — CHANGES_REQUESTED** [review]
Inconsequential bits, otherwise LGTM

**Review by silasary — APPROVED** [review]

**Pull request merged**